.github/shared/src/swagger.js

```diff
@@ -177,3 +266,207 @@ function json(file) {
+        if (["stable", "preview"].some((v) => v === s[1])) {
+          const version = segments[s[0] + 1];
+          if (version) {
+            return apiVersionRegex.test(version) && !withPreview
+              ? version.substring(0, 10)
+              : version;
+          }
+        }
+      }
+    }
+  } else {
+    if (segments && segments.length > 1) {
+      for (const s of segments) {
+        if (apiVersionRegex.test(s)) {
+          return withPreview ? s : s.substring(0, 10);
+        }
+      }
+    }
+  }
+
+  try {
+    // Check if file exists using access()
+    await access(filePath);
+    const content = await readFile(filePath, "utf8");
+    return JSON.parse(content)?.info?.version;
+  } catch {
+    // File doesn't exist or can't be read/parsed
+    // This is expected for non-existent files, so we don't log a warning
+  }
+
+  return undefined;
+}
+
+/**
+ * Get preceding swagger version for a specific lifecycle stage
+ * @param {string} targetSwaggerPath - Path to the swagger file to find predecessors for
+ * @param {Swagger[]} availableSwaggers - Array of swagger objects to search within
+ * @param {string} versionKind - Version lifecycle stage (PREVIEW or STABLE)
+ * @returns {Promise<string | undefined>} - Path to preceding version or undefined if not found
+ */
+async function getPrecedingSwaggerByType(targetSwaggerPath, availableSwaggers, versionKind) {
+  const targetSwagger = availableSwaggers.find((s) => s.path === targetSwaggerPath);
+  if (!targetSwagger) {
+    return undefined;
+  }
+
+  const currentVersion = await targetSwagger.getVersion();
+  const fileName = await targetSwagger.getFileName();
+
+  try {
+    // Load version info for all swaggers to enable filtering
+    const swaggersWithVersions = await Promise.all(
+      availableSwaggers.map(async (swagger) => ({
+        swagger,
+        version: await swagger.getVersion(),
+        fileName: await swagger.getFileName(),
+        versionKind: swagger.versionKind,
+      })),
+    );
+
+    const versionsOfType = swaggersWithVersions.filter(
+      (item) =>
+        item.fileName === fileName &&
+        item.versionKind === versionKind &&
+        item.version <= currentVersion &&
+        item.swagger.path !== targetSwaggerPath,
+    );
+
+    if (versionsOfType.length > 0) {
+      const mostRecent = versionsOfType.reduce((previous, current) =>
+        previous.version > current.version ? previous : current,
+      );
+      return mostRecent.swagger.path;
+    }
+  } catch {
+    // Version not found or error occurred
+  }
+
+  return undefined;
+}
+
+/**
+ * Get operations from all previous versions that also exist in current swagger
+ * @param {string} targetSwaggerPath - Path to the current swagger file
+ * @param {Swagger[]} availableSwaggers - Array of swagger objects to search within
+ * @returns {Promise<LegacyOperation[]>} - Operations that exist in previous versions which exist in current version too
```

Comment:
```suggestion
 * @returns {Promise<Map<string, LegacyOperation[]>} - Operations that exist in previous versions which exist in current version too
```

Reply:
Now it returns a map with key as the swagger path and value as the matched operations from this swagger.